### PR TITLE
MINOR: Use thread name and task for sensor name

### PR DESCRIPTION
**Pull request by bbejeck**
Changes to keep the operation name as is and make the sensor name unique.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
@guozhangwang @mjsax @vvcephei

**Comment by mjsax** [comment]
This is a follow up to #5009 that incorrectly changes not only sensor names, but also metric names that should not change.

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by vvcephei** [comment]
What do you think about this instead (so we can avoid mutable `taskName` in this multithreaded class):
```
diff --git a/streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java b/streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
index bc2e15012..6a3b22cc2 100644
--- a/streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
+++ b/streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java
@@ -253,12 +253,30 @@ public class StreamsMetricsImpl implements StreamsMetrics {
                                                 final String operationName,
                                                 final Sensor.RecordingLevel recordingLevel,
                                                 final String... tags) {
-        return addLatencyAndThroughputSensor(
-            scopeName,
-            entityName,
-            threadName + "." + taskName + "." + operationName,
+
+        final Map<String, String> tagMap = constructTags(scopeName, entityName, tags);
+        final Map<String, String> allTagMap = constructTags(scopeName, "all", tags);
+
+        // first add the global operation metrics if not yet, with the global tags only
+        final Sensor parent = metrics.sensor(
+            sensorName(threadName + "." + taskName + "." + operationName, null),
+            recordingLevel
+        );
+        addLatencyMetrics(scopeName, parent, operationName, allTagMap);
+        addThroughputMetrics(scopeName, parent, operationName, allTagMap);
+
+        // add the operation metrics with additional tags
+        final Sensor sensor = metrics.sensor(
+            sensorName(threadName + "." + taskName + "." + operationName, entityName),
             recordingLevel,
-            tags);
+            parent
+        );
+        addLatencyMetrics(scopeName, sensor, operationName, tagMap);
+        addThroughputMetrics(scopeName, sensor, operationName, tagMap);
+
+        parentSensors.put(sensor, parent);
+
+        return sensor;
     }
 
     /**
@@ -292,13 +310,27 @@ public class StreamsMetricsImpl implements StreamsMetrics {
                                       final String operationName,
                                       final Sensor.RecordingLevel recordingLevel,
                                       final String... tags) {
-        return addThroughputSensor(
-            scopeName,
-            entityName,
-            threadName + "." + taskName + "." + operationName,
+        final Map<String, String> tagMap = constructTags(scopeName, entityName, tags);
+        final Map<String, String> allTagMap = constructTags(scopeName, "all", tags);
+
+        // first add the global operation metrics if not yet, with the global tags only
+        final Sensor parent = metrics.sensor(
+            sensorName(threadName + "." + taskName + "." + operationName, null),
+            recordingLevel
+        );
+        addThroughputMetrics(scopeName, parent, operationName, allTagMap);
+
+        // add the operation metrics with additional tags
+        final Sensor sensor = metrics.sensor(
+            sensorName(threadName + "." + taskName + "." + operationName, entityName),
             recordingLevel,
-            tags
+            parent
         );
+        addThroughputMetrics(scopeName, sensor, operationName, tagMap);
+
+        parentSensors.put(sensor, parent);
+
+        return sensor;
     }
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java [hunk 1]
```diff
@@ -118,24 +118,27 @@ public void testMetrics() {
         assertNotNull(metrics.getSensor("name-mock.0_0." + throughputOperation));
 
         for (final String opName : latencyOperations) {
-            StreamsTestUtils.getMetricByNameFilterByTags(metrics.metrics(), "mock.0_0." + opName + "-latency-avg", groupName, metricTags);
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:121]: L116 and 118 also have changes that should be reverted.

**Review by vvcephei — CHANGES_REQUESTED** [review]
Thanks for the patch, @bbejeck.

I'm concerned about the unprotected mutable state this introduces. If multithreaded access to the methods in question occurs, then the task name may be overwritten by the second caller before the first caller creates the sensor name, resulting in incorrect execution.

Also, there is one place where we need to revert a couple more lines from #5009 to return the tests to their original state.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:121]: hmm, I'm not sure about that.  Those lines are getting the sensor by name which still uses the combination of `thread.name() + taskID + operationName` or am I missing something?

**Comment by bbejeck** [comment]
updated this.  

@mjsax @vvcephei  have another look and let me know your thoughts on this change.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java [hunk 2]
```diff
@@ -229,16 +229,33 @@ public Sensor addLatencyAndThroughputSensor(final String scopeName,
                                                 final String operationName,
                                                 final Sensor.RecordingLevel recordingLevel,
                                                 final String... tags) {
+
+        return addThroughputSensor(null,
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: I think you meant `addLatencyAndThroughputSensor`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:121]: oh! yeah, you're right. Sorry!

**Review by vvcephei — APPROVED** [review]
There's one typo, otherwise LGTM! Thanks again.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java]: doh! yes, good catch

**Review by guozhangwang — APPROVED** [review]
LGTM! Thanks @bbejeck

**Pull request merged**